### ISSUE TITLE
GET /v2/registrations/<registrationId> 

### ISSUE THREAD

**Issue by jmcanterafonseca** [telefonicaid/fiware-orion#3008]
Allow to obtain information about a particular registration.

**Comment by fgalan** [comment]
(I'll provide some hints regarding data model, documentation and implementation, each one in a separated comment. I'll start by data model)

_DB data model considerations_

In the same way NGSIv1 subscriptions were integrated with NGSIv2 subscriptions from a DB data model model point of view, we are going to do the same for registrations. Thus, let's analyze side by side the API payload (from NGSIv2 spec):

```
      {
            "id": "abcdefg",
            "description": "Example Context Source",
            "dataProvided": {
              "entities": [
                {
                  "id": "Bcn_Welt",
                  "type": "Room"
                }
              ],
              "attrs": [
                "temperature"
              ]
            },
            "provider": {
              "http": {
                "url": "http://contextsource.example.org"
              },
              "supportedForwardingMode": "all",
              "legacyForwarding": true
            },
            "expires": "2017-10-31T12:00:00",
            "status": "failed",
            "forwardingInformation": {
              "timesSent": 12,
              "lastForwarding": "2017-10-06T16:00:00.00Z",
              "lastFailure": "2017-10-06T16:00:00.00Z",
              "lastSuccess": "2017-10-05T18:25:00.00Z",
            }
} 
```

(actually, `legacyForwarding` is not in the NGSIv2 spec, but it is part of the particular format implemented by Orion).

and the current data model (see https://fiware-orion.readthedocs.io/en/master/admin/database_model/index.html#registrations-collection):

```
 {
   "_id": ObjectId("5149f60cf0075f2fabca43da"),
   "fwdRegId": "5149f60cf0075f241bca22f1",
   "expiration": 1360232760,
   "contextRegistration": [
       {
           "entities": [
               {
                   "id": "E1",
                   "type": "T1",
                   "isPattern": "false"
               },
               {
                   "id": "E2",
                   "type": "T2",
                   "isPattern": "false"
               }
           ],
           "attrs": [
               {
                   "name": "A1",
                   "type": "TA1",
                   "isDomain": "false"
               },
               {
                   "name": "A2",
                  "type": "TA2",
                   "isDomain": "true"
               }
           ],
           "providingApplication": "http://foo.bar/notif"
      },
  ]
 }
```

Thus, lest's analyze how to map each field in the API payload to DB:

* `id` maps directly to `_id`
* `description` doesn't maps with any existing field. A new field has to be defined for it. In abscense of the field in the DB (i.e. GET on registrations created by NGSIv1) this field is not shown in the payload (as happens with subscriptions, as far as I remember).
* `entities` and `attrs` within `dataProvided` maps to `contextRegistrations`. Note that `contextRegistrations` is actualy a list, so what we would have for registrations created with the NGSIv2 API is a mono-item `contextRegistrations`.
  * `entities` will use `id`, `isPattern`, `type` and `typePattern` in the same way currently csubs collection uses (`isPattern` as a string, `typePattern` as a bool).
  * `attrs` elements have `type` and `isDomain` in the DB model. Niether `type` and `isDomain` is needed in the NSGIv2 registrations API, so they can be filled blank (in the case of type) and `"false"` (in the case of `isDomain`) at DB.
* `provider.http.url` maps directly to `providingApplication`
* `provider.supportedForwardingMode` is not supported by the moment. This doesn't have a correspondence in DB and will be rendered as `all` always.
* With regards to `provier.legacyForwarding`, this fields doesn't need yet a field in the database, as all the registrations will use legacy forwarding (in fact, NGSIv2 forwarding has not been yet defined in the spec). Thus, as with `supportedForwardingMode`, this will be always rendered as `true`.
* `expires` maps directly with `expiration`
* `status` doesn't maps with any existing field. A new field has to be defined for it. In abscense of the field (i.e. GET on registrations created by NGSIv1) "active" has to be rendered.
* The subfields in `forwardingInformation` hasn't a direct correspondence with any existing field. Four fields will be created for them, using the same names that the similar ones in csubs collection: `count`, `lastNoficiation`, `lastSucess` and `lastFailure`.

**Comment by fgalan** [comment]
Caveat: how to render a registration with N elements in `contextRegistrations` (because it was created with a NGSIv1 registration) given that the NGSIv2 payload has only one `dataProvided`?

The simple solution is to "collapse" all the contextRegistrations elements, i.e. putting together all the `contextRegistrations.entities` list and putting toggether all the `contextRegistrations.attrs` lists, then render as one mono-contextRegistrations way.

E.g. (simplified notation):

```
contextRegistraitons = [
  entities = [E1, E2] : attrs = [A1, A2],
  entities = [E3] :  attrs = [A3, A4],
  entities = [E4] :  attrs = [A5, A6]
]
```

will render as:

```
dataProvided = {
  entities = [E1, E2, E3, E4] : attrs = [A1, A2, A3, A4, A5, A6]
}
```

**Comment by fgalan** [comment]
_Documentation considerations_

Most of the documentation is already in the NGSIv2 specification. The only pending work would be some aspects regarding the particular implementation in Orion in its .md files. In particular, a new section has to be added to https://fiware-orion.readthedocs.io/en/master/user/ngsiv2_implementation_notes (suggested name "Registration") explaining that:

* The `legacyForwarding` field, stating that by the moment this is the only way supported by the implementation.
*  Only supported `supportedForwardingMode` mode is `all`.
* `dataProvided.expression` is not yet supported
* `inactive` status is not yet supported.

In addition, the following section should be reviewed: https://fiware-orion.readthedocs.io/en/master/user/v1_v2_coexistence/index.html#ngsiv1-notification-with-ngsiv2-subscriptions. At least the following sentence may become obsolete: 

> Context availability management functionality (i.e. operations to register Context Providers) is still to be implemented for NGSIv2

**Comment by fgalan** [comment]
_Implementation notes_

As with any other new operation in Orion, we have to take into consideration the usual three layers:

1. Service routines
2. Parsing
3. Mongo backend

With regards with service routines, nothing special to do. This is just any other GET operation. With regards to parsing, being a GET operation we don't need parsing this time. Let's focus so in mongoBackend.

I'd suggest to follow the same strategy we used with subscriptions, which at the end has been very successful to implement the "NGSIv1 to NGSIv2 handover" in the code. That is, to create a new a "fresh" structure of operations and associated classes intead of trying to reuse the existing classes.

In this sense we need:

* A `Registration` class, defined in `apiTypesV2/Registartion.[h|cpp]`. It will be similar to existing `apiTypesV2/Subscription.[h|cpp]` and written in the same style.
* A `mongoBackend/mongoGetRegistration.[h|cpp]` implementation. It will be similar to existing `mongoBackend/mongoGetSubscription.[h|cpp]` and written in the same style, e.g. `set*()` method to field each element.

**Comment by fgalan** [comment]
_Test considerations_

I don't foresee unit test for this functionality.

With regards to functional test, just creat some registrations using `POST /v1/registerContext` (e.g. with one context registration and with severael context registrations), then issue `GET /v2/registrations/<id>` to check it works as expected.

In the future, these .test could be modified in order to replace `POST /v1/registerContext` by `POST /v2/registrations`. A FIXME note in the .test could be included in order not forgetting that adjustment.

**Comment by fgalan** [comment]
Documentation considerations (2)

Some additional elements, that I misses in my [previous comment on documentation](https://github.com/telefonicaid/fiware-orion/issues/3008#issuecomment-336490272):

* The https://fiware-orion.readthedocs.io/en/master/user/walkthrough_apiv2/index.html document should be expanded with context availaibiltiy managemnet sections, in a similar way these sections are in the v1 version of the walkthrough
* Overal review to the user manual in order to expand context availability related documentation with NGSIv2 information. Hint: use `grep *.md 'v1/registerContext' would make the trick`
* Devel documentation at https://github.com/telefonicaid/fiware-orion/tree/master/doc/manuals/devel should be expansed with the new flows corresponding to NGSIv2 context availability management. At least five new MB-XX flows must be added.

**Comment by fgalan** [comment]
All the registration features share the same common base branch: *feature/ngsiv2_registrations*.

Thus, this issue should be implemented in a branch splited from feature/ngsiv2_regsitrations (instead of master) and the PR done using feature/ngsiv2_registrations as base branch (instead of master).

feature/ngsiv2_registrations will be update regularly from master, to keep it as sync as possible and ease the final feature/ngsiv2_registrations -> master merge once all the NGSIv2 registrations management  functionality gets completed.

**Comment by kzangeli** [comment]
About *"how to render a registration with N elements"* ...

Are we sure we want to do it this way?
If I have understood this correctly, responding in that "collapsed way", the response does not reflect the truth about the registration.

I would prefer to add an optional URI param `?respondWithRegistrationVector=true` (or similar :-)), and when asked for a `GET /v2/reg/xxx`, that has more than one contextRegistration, we'd return an error if that URI param is not present (as we're unable to answer correctly).
And if present, respond with a vector instead of a single element (of Registration).

Not perfect, having two different outgoing payloads for the same service (almost the same service, as the URI parameter alters it), but **much** better than to lie to the user.
Actually, not the first time that URI parameters would alter the response, we have plenty of examples of that.
Just the first time the **JSON type** of the response would be altered.

[ If the URI param is present, the response would always be a vector, even when not necessary ]

**Comment by fgalan** [comment]
> About "how to render a registration with N elements" ...
>
> Are we sure we want to do it this way?

Being a secondary aspect, let's move out this of this issue by the moment. I have created a separate issue for that https://github.com/telefonicaid/fiware-orion/issues/3044.

In the context of the present issue assume that registration will not have more than one CR elements in the vector.

**Comment by kzangeli** [comment]
The *service path* of the registration is missing from the 'response examples'.
For a GET we should probably respond with **all** the registration, right?

**Comment by fgalan** [comment]
Another item for the "implementation notes" for NGSIv2: `idPattern` and `typePattern` are not supported (same than in NGSIv1)

**Comment by fgalan** [comment]
The "core" of the implementation has been done in PR https://github.com/telefonicaid/fiware-orion/pull/3060

**Comment by fgalan** [comment]
The basic functionality is completed. Advance functionality has been moved to spin off issues:

* Multi-CRs comming from NGSIv1 registrations: #3044 
* SupportedForwardingModde support: #3106 
* Expressions support: #3107
* Support inactive status: #3108
* legacyForwardingMode different from true (actually, related with NGSIv2 forwarding):

The issues is thus closeable once **feature/ngsiv2_registrations lands on master**. Assigned to Orion 1.12.0, as this is probably the version in which this landing will take place.
